### PR TITLE
Add NUOPC prototype demonstrating coupling a Julia-based component

### PR DESCRIPTION
**Pull request by billsacks**
The new prototype in this PR demonstrates the ability to couple a component written in the Julia language. The NUOPC cap is written in C (and so the prototype builds on the ESMX_SingleModelInCProto that @theurich put together), with calls from the C layer to the Julia component, as well as calls from Julia back to ESMF (via a thin wrapper in C).

See the README.md file for this prototype for an explanation of some design choices and implementation details.

Note that this new prototype will only be run in nightly testing on machines where the ESMF_TEST_NUOPC_JULIA environment variable is set to ON - so that we won't try to run it on systems that don't have Julia installed. I have turned this on for my Mac:

https://github.com/esmf-org/esmf-test-scripts/blob/baede047b3b2de6d09b2df994f1cdf63b881158c/config/green.yaml#L29

and I plan to do the same on derecho, for at least one combo (by adding a 'module load julia'); I have tested this manually on derecho and have confirmed that it passes there.

### CODE REVIEW

**Comment by billsacks** [comment]
@theurich and @danrosen25 - I'm assigning you both as reviewers. I don't feel like this absolutely needs a review, but I'd be happy to get at least a quick review from one of you if you have the time and interest.

**Comment by billsacks** [comment]
@danrosen25 had the good suggestion of adding some passing back and forth of Fields, exchanging data with an ESMX Data component. I'd like to add that. This will partly build on some additional pieces from `ESMX_AtmOcnFortranAndCProto` – adding some pieces to the C cap – and then will add to that some extra hand-offs between Julia and C. A lot of this will be about demonstrating Julia-C interoperability rather than actually demonstrating features of NUOPC (since the features of NUOPC that this demonstrates probably won't go much, if at all, beyond what's in `ESMX_AtmOcnFortranAndCProto`), but it still seems like this could be a helpful step towards demonstrating a prototype Julia component.

My thinking, though, is that we should bring in this initial version without that addition, and then I can do another iteration that adds those pieces.

**Comment by billsacks** [comment]
@danrosen25 asked how I installed Julia on my Mac. I can't find notes on this, but I have a feeling I just used the instructions here:

https://julialang.org/downloads/

It looks like that automatically(?) added some things to my dot files to put julia in my path.

**Comment by mochell** [comment]
I think I at least understand what is happening, its great!
Once you come to the iteration to exchange fields, I could try to provide init() and run() that wrap around the wave model. Or you just demonstrate how to hold the state variables in Julia (as @billsacks did already in the other dummy code).
I think what needs to happen first is how to maintain the correct Julia version environment though. PiCLES will be released as a package in a few months from now, loading a predefined manifest.toml should do it then.
For exchanging the fields we would have to decide if we jump to CESM-like fields and grids right away, or go with some dummy fields first. I think the exchanged fields should be analogous to what Paul Hall is doing with WW3.
I probably need to understand what ESMX_AtmOcnFortranAndCProto is doing..

**Comment by billsacks** [comment]
Thanks @mochell ! Let's talk more about next steps to prototype the field exchange. I'll reply via email to set up a time.

**Comment by billsacks** [comment]
FYI, I have enabled testing this Julia prototype in one derecho combo (https://github.com/esmf-org/esmf-test-scripts/commit/7b9128d107b7d9588eac892a3c810c78f6121611). I have tested this manually on this branch to ensure it passes. It will become active once this PR is merged to develop.

**Comment by navidcy** [comment]
interesting!

File: ESMX_SingleModelInJuliaProto/SiMoCoJulia/CMakeLists.txt
Comment (danrosen25) [inline ESMX_SingleModelInJuliaProto/SiMoCoJulia/CMakeLists.txt]: The NUOPC C interface wasn't added until after 8.6.0 (and also isn't in the patch 8.6.1)

**Review by danrosen25 — CHANGES_REQUESTED** [review]

File: ESMX_SingleModelInJuliaProto/SiMoCoJulia/CMakeLists.txt
Comment (billsacks) [reply ESMX_SingleModelInJuliaProto/SiMoCoJulia/CMakeLists.txt]: Very good point. I just pushed a fix.

File: ESMX_SingleModelInJuliaProto/Makefile
Comment (danrosen25) [inline ESMX_SingleModelInJuliaProto/Makefile:34]: This is the type of thing I'd like the ESMX build infrastructure to handle, if possible. I unsuccessfully tried to put it into the `esmxBuildDL.yaml` file. I won't have time to work on this.

**Review by danrosen25 — COMMENTED** [review]
Overall I think this simple model is working fine but the build process and linking is confusing. If it can be cleaned up so that `ESMX_Builder esmxBuildDL.yaml` can be run without any prerequisite steps (i.e. setting LDFLAGS) that will simplify the experience for the model developer. I am hoping this can be done through extensions to ESMX and a FindJulia.cmake file that provides all the information needed.

**Review by danrosen25 — COMMENTED** [review]
Overall I think this simple model is working fine but the build process and linking is confusing. If it can be cleaned up so that `ESMX_Builder esmxBuildDL.yaml` can be run without any prerequisite steps (i.e. setting LDFLAGS) that will simplify the experience for the model developer. I am hoping this can be done through extensions to ESMX and a FindJulia.cmake file that provides all the information needed.

**Comment by billsacks** [comment]
@danrosen25 thanks a lot for your thoughts on this. I actually first started down a path of trying to put in more ESMX build features to better support this, but gave up. The path I went down differed from the one I think you're suggesting, but I'll share my thoughts here and see what you think about the best path forward.

First, the rationale I had for giving up on this path and going with the approach here (which is not as elegant as I ideally wanted) was that I felt like the use of ESMX here was more incidental rather than being central to this prototype. ESMX was a huge help in getting this up and going easily - and I was very grateful for that! - but I didn't necessarily see a use case for users wanting to use ESMX with a Julia model themselves at this point, so it felt like it might not make sense to spend a lot of time trying to get that build piece working more elegantly - since it wouldn't extend to someone coupling a Julia component to their own system (like CESM) outside of the ESMX context. So I came to feel like it would be best to get something working without investing too much time and without complicating ESMX for this purpose. That said, now that I write this out, I guess I see how the developer of the Julia component for CESM might like to first get things working with ESMX before going the next step towards CMEPS / CESM coupling.

The path I initially took was a bit of a hybrid between what's currently here and what I think you might be thinking: I tried adding a mechanism to pass extra link libraries into the ESMX build via the ESMX_Builder rather than using an environment variable:

```diff
diff --git a/src/addon/ESMX/CMakeLists.txt b/src/addon/ESMX/CMakeLists.txt
index 4d5192427d..40cf6fc0ff 100644
--- a/src/addon/ESMX/CMakeLists.txt
+++ b/src/addon/ESMX/CMakeLists.txt
@@ -45,6 +45,9 @@ endif()
 add_executable(${ESMX_EXE_NAME} ESMX_App.F90)
 target_include_directories(${ESMX_EXE_NAME} PUBLIC ${PROJECT_BINARY_DIR})
 target_link_libraries(${ESMX_EXE_NAME} PUBLIC esmx_driver)
+if (DEFINED ESMX_EXTRA_LINK_LIBS)
+   target_link_libraries(${ESMX_EXE_NAME} PUBLIC ${ESMX_EXTRA_LINK_LIBS})
+endif()
 set_target_properties(${ESMX_EXE_NAME} PROPERTIES INSTALL_RPATH_USE_LINK_PATH TRUE)
 
 # Install executable
diff --git a/src/apps/ESMX_Builder/ESMX_Builder.sh b/src/apps/ESMX_Builder/ESMX_Builder.sh
index 56e994f273..e18a9a214e 100755
--- a/src/apps/ESMX_Builder/ESMX_Builder.sh
+++ b/src/apps/ESMX_Builder/ESMX_Builder.sh
@@ -30,6 +30,8 @@ usage () {
   printf "      set --build-type=debug\n"
   printf "  --build-args=BUILD_ARGS\n"
   printf "      cmake arguments (e.g. -DARG=VAL)\n"
+  printf "  --extra-link-libs=EXTRA_LINK_LIBS\n"
+  printf "      extra libraries to link against when building the ESMX executable\n"
   printf "  --disable-comps=DISABLE_COMPS\n"
   printf "      disable components\n"
   printf "  --build-jobs=BUILD_JOBS\n"
@@ -63,6 +65,7 @@ settings () {
   printf "  INSTALL_PREFIX=${INSTALL_PREFIX}\n"
   printf "  BUILD_TYPE=${BUILD_TYPE}\n"
   printf "  BUILD_ARGS=${BUILD_ARGS}\n"
+  printf "  EXTRA_LINK_LIBS=${EXTRA_LINK_LIBS}\n"
   printf "  DISABLE_COMPS=${DISABLE_COMPS}\n"
   printf "  BUILD_JOBS=${BUILD_JOBS}\n"
   printf "  MODULEFILE=${MODULEFILE}\n"
@@ -80,6 +83,7 @@ BUILD_FILE="esmxBuild.yaml"
 ESMF_ESMXDIR=""
 BUILD_TYPE="release"
 BUILD_ARGS=""
+EXTRA_LINK_LIBS=""
 DISABLE_COMPS=""
 BUILD_JOBS=""
 BUILD_DIR="${CWD}/build"
@@ -122,6 +126,9 @@ while [[ $# -gt 0 ]]; do
     --build-args=?*) BUILD_ARGS=${1#*=} ;;
     --build-args)  usage_error "$1" "requires an argument" ;;
     --build-args=) usage_error "$1" "requires an argument" ;;
+    --extra-link-libs=?*) EXTRA_LINK_LIBS=${1#*=} ;;
+    --extra-link-libs)  usage_error "$1" "requires an argument" ;;
+    --extra-link-libs=) usage_error "$1" "requires an argument" ;;
     --build-jobs=?*) BUILD_JOBS=${1#*=} ;;
     --build-jobs)  usage_error "$1" "requires an argument" ;;
     --build-jobs=) usage_error "$1" "requires an argument" ;;
@@ -233,6 +240,9 @@ fi
 if [ ! -z "${BUILD_ARGS}" ]; then
   CMAKE_SETTINGS+=("-DESMX_BUILD_ARGS=${BUILD_ARGS}")
 fi
+if [ ! -z "${EXTRA_LINK_LIBS}" ]; then
+  CMAKE_SETTINGS+=("-DESMX_EXTRA_LINK_LIBS=${EXTRA_LINK_LIBS}")
+fi
 
 # make settings
 BUILD_SETTINGS=("")
```

I followed the example of the `--build-args` argument to ESMX_Builder, but this wasn't working for me: for some reason it wasn't passing through the extra link libs to the CMake link line. I also experimented with the existing `--build-args` flag but couldn't get that working either: it seemed like the things I specified there weren't actually getting passed through to the CMake build. I didn't spend too much time on this before giving up and going what felt like the current, simpler approach. This is partly because the other issue I was having with the ESMX_Builder approach was that it felt like it was going to be messy to specify multiple extra link libraries, due to the way shell quoting was happening; I don't actually need that in my use case, but I felt uncomfortable having something that only supported a single link lib and thought it would take some time to figure out how to support multiple.

But: If I understand your thinking, I think you may be imagining adding something more Julia-specific to the ESMX build. I think that might actually be (significantly?) easier, if that feels like the right path forward to you: I think I could copy the FindJulia.cmake from this PR into the ESMX cmake area, then add a flag like `--enable-julia` to the ESMX_Builder (or maybe an option in the yaml build configuration?), which, if set, will do something like the current Julia stuff in the SiMoCoJulia CMakeLists.txt. Would you support this approach of adding some Julia-specific things to the ESMX build? If so, I can try that... but it would probably mean this won't make it into 8.7, because I probably can't get to this until next week. (That's okay with me if it is with you.)

**Comment by danrosen25** [comment]
https://github.com/danrosen25/nuopc-app-prototypes/tree/genio/ESMX_GenericIOProto

**Comment by billsacks** [comment]
From core call today: @danrosen25 doesn't want Julia-specific stuff in ESMX but  suggests leveraging capabilities in ESMX for flexibly adding libraries.

Dan points to https://github.com/esmf-org/esmf/blob/develop/src/addon/ESMX/Driver/CMakeLists.txt#L124 (though says there may be a bug in there)

Then in esmxBuildDL.yaml, in the application setting, add something like this:
- `link_module_paths: SiMoCoJulia/cmake`
- `link_packages: Julia`
- `link_libraries: ${Julia_LIBRARY}`

But this isn't working, because it expands ${Julia_LIBRARY} before the find_package call for Julia, so Julia_LIBRARY isn't set yet.

One way to do this would be to add something similar to link_libraries but that specifies the 

A workaround that got this working in @danrosen25 's experimentation is:
- process the dynamically-created file (appConf.txt) twice... the cleaner thing would be to split that into two files, one containing the settings needed for find_package, and the other with setting variables that should be done after the find_package calls
- don't try to do find_library on something that's already a full path, or maybe add hints to the find_library call, or something similar (e.g., testing if the specified library exists with `if (EXISTS ${ESMX_LINK_LIBRARY})` or using `cmake_path(IS_ABSOLUTE <path-var> <out-var>)`, and if so, just do `target_link_libraries` without trying to do the find_library) (Dan came up with a solution to this).

Dan will create a branch with the necessary changes to ESMX. With his changes, the above lines to esmxBuildDL.yaml allow things to work with the LDFLAGS stuff removed from the Makefile.

**Comment by billsacks** [comment]
@danrosen25 approved merging this as is, and then we'll clean up the build as a follow-up.

**Comment by danrosen25** [comment]
Follow-up: The issues preventing Julia linking within ESMX have been resolved [here](https://github.com/danrosen25/esmf/tree/fix/esmx_app_link). There's an open issue to merge these changes [here](https://github.com/esmf-org/esmf/issues/281). These fixes are compatible with the following esmxBuild.yaml configuration. Note that Julia_LIBRARY is a CMake variable provided by the FindJulia.cmake module when called using find_package(Julia), which happens during the `link_packages` step of the ESMX build.
```
link_module_paths: SiMoCoJulia/cmake
link_packages: Julia
link_libraries: ${Julia_LIBRARY}
```

**Pull request merged**
